objects/isn_unifiedgrowingtray.lua

```diff
@@ -0,0 +1,541 @@
+		local newGrowth = storage.growth + deltaGrowth
+		if newGrowth > storage.stage[storage.currentStage].val then
+			local refundTime
+			local growthDif
+			local failToStages = false
+			if newGrowth > storage.growthCap then
+				local stageDif = storage.stages - storage.currentStage
+				growthDif = newGrowth - storage.growthCap
+				local hasFluid = true
+				if stageDif > 0 then hasFluid = isn_doFluidConsume(stageDif) end
+				if not hasFluid then failToStages = true end
+			else
+				failToStages = true
+			end
+			if failToStages then
+				growthDif = newGrowth - storage.stage[storage.currentStage].val
+			end
+			if storage.ffCycles < self.maxFFCycles then
+				storage.ffCycles = storage.ffCycles + 1
+				refundTime = (growthDif / deltaGrowth) * secondsThisUpdate
+			else
+				storage.ffCycles = 0
+				refundTime = 0
+			end
+			storage.lastWorldTime = storage.lastWorldTime - refundTime
+			newGrowth = newGrowth - growthDif
+		end
+		storage.growth = newGrowth
+	end
+	
+	--Check if we should try to use more fluid due to growth. (DON'T check the last stage)
+	if storage.currentStage < storage.stages then
+		if storage.growth >= storage.stage[storage.currentStage].val then
+			storage.hasFluid = isn_doFluidConsume()
+			storage.currentStage = storage.currentStage + 1
+		end
+	end
+	
+	--If the above fluid use left us dry stop here (don't output items) and try again later.
+	--Yes this may generate 1 more (unnecessary) growth spurt but it won't use more fluid.
+	if not storage.hasFluid then return end
+	
+	--Check if we are grown.
+	if storage.growth >= storage.growthCap then
+		--Spawn our output...
+		for i = 1, storage.yield do
+			--Generate the items.
+			local itemDescriptors = nil
+			if storage.hasTree then
+				itemDescriptors = isn_genSaplingDrops()
+			elseif root.isTreasurePool(storage.harvestPool) then
+				itemDescriptors = root.createTreasure(storage.harvestPool, world.threatLevel())
+			end
+			--'Drop' the items.
+			if type(itemDescriptors) == "table" then
+				for _, itemDescriptor in pairs(itemDescriptors) do
+					local mask = {0, 1, 2}
+					if itemDescriptor.name == storage.currentseed.name then
+						mask = {1, 2}
+					end
+					fu_sendOrStoreItems(0, itemDescriptor, mask)
+					--next line might not be necessary... was in original growingtray but not hydroponicstray code.
+					--world.containerAddItems(entity.id(), itemDescriptor)
+				end
+			end
+		end
+		--Recycle the plant...
+		isn_doRecyclePlant()
+	end
+	
+end
+
+--Updates internal fluid levels and modifies a growth variable depending on
+--liquid in slot.
+--optional arg fluidNeed is amount of fluid required to top up to.
+function isn_doWaterIntake(fluidNeed)
+	local contents = world.containerItems(entity.id())
+	local water = contents[storage.waterslot]
+	if not water then return false end
+	
+	for key, value in pairs(config.getParameter("isn_waterInputs")) do
+		if water.name == key and type(value) == "table" then
```

Comment:
you can replace those lines with something like : 
```lua
	local watertable = config.getParameter("isn_waterInputs")
	if watertable[water.name] and type(watertable[water.name]) == "table" then
```
notes: you need to change the variable name I came up with.
note2: this table need to be always loaded in init imho because we're not gonna reopen it each & everytime
note3: again too many sanity check, is there a valid reason why type(watertable[water.name]) (or type(value) == "table" in your original code) wouldn't be a table ??? & before you ask, no, modder error is not a valid reason.

Reply:
RE the key lookup... <forehead> yeah, that was the structure of table lookups in the original code (both growingtable.lua and hydroponicstray.lua)... I didn't even think about changing it to the more efficient key lookup.  Good point on just caching the table too.  Only issue I see is it could increase the memory load some since EACH tray in the game will be carrying around a copy of those tables (depends on how things work under the hood).  In the context of a server with a bunch of people with huge bases that extra memory load could be an issue.  Probably not a big issue right now as the tables are so very short but then look at how the old seed->output table grew when trying to add support for more mods.

A global (shared between all trays) would work better.  Something in init vaguely like:
`if not globalTableCache then
    --read the table from config
end`

Actually something like that but storing all configuration data could be employed so that we only have 1 init check to load the data.  I'm just cautious about how globals work in this context.  More reading and testing (for either case).

RE sanity checks:  It was less my concern about the ability of the FU modders and more if someone does a patch file with horrible values.  But I'm fine assuming expert ability for most values.

Not sure I completely follow note 3...  Did you ask a question about my logic using your code and logic (with a name that you told me not to use in note 1)?

(guessing from previous context)
I think it might be a really old sanity check.  In general some of the sanity checks that are in there are older and from when I was testing things and probing what Lua (and Starbound) could do.

(In general)
Some sanity checks are cheap, either rarely run, or relative to what is lost if something bad gets through.  I find some of them make the code easier to read.  I'm loath to optimise the code so much that only I or an actual expert Lua coder can follow without spending a ton of time trying to follow what is expected.